tools/certs/Makefile

```diff
@@ -277,3 +279,93 @@ root-key.pem:
 		-extensions req_ext -extfile $(dir $<)/workload.conf \
 		-in $< -out $@
 
+#------------------------------------------------------------------------
+
+.PHONY: %-certs-vm
+
+%-certs-vm: %/vm-workload-cert.pem
+	@make --makefile $(realpath $(firstword $(MAKEFILE_LIST))) clean
+	@echo "done"
+
+%/vm-workload-cert.pem: %/vm-workload.csr  k8s-root-cert.pem k8s-root-key.pem
+	@echo "generating $@"
+	@openssl x509 -req -days $(WORKLOAD_DAYS) \
+		-CA k8s-root-cert.pem  -CAkey k8s-root-key.pem -CAcreateserial\
+		-extensions req_ext -extfile $(dir $<)/vm-workload.conf \
+		-in $< -out $@
+%/vm-workload.csr: L=$(dir $@)
+%/vm-workload.csr: %/key.pem %/vm-workload.conf
+	@echo "generating $@"
+	@openssl req -new -config $(L)/vm-workload.conf -key $< -out $@
+
+%/vm-workload.conf: L=$(dir $@)
+%/vm-workload.conf:
+	@echo "[ req ]" > $@
+	@echo "encrypt_key = no" >> $@
+	@echo "prompt = no" >> $@
+	@echo "utf8 = yes" >> $@
+	@echo "default_md = sha256" >> $@
+	@echo "default_bits = $(INTERMEDIATE_KEYSZ)" >> $@
+	@echo "req_extensions = req_ext" >> $@
+	@echo "x509_extensions = req_ext" >> $@
+	@echo "distinguished_name = req_dn" >> $@
+	@echo "[ req_ext ]" >> $@
+	@echo "subjectKeyIdentifier = hash" >> $@
+	@echo "basicConstraints = critical, CA:false" >> $@
+	@echo "keyUsage = digitalSignature, keyEncipherment" >> $@
+	@echo "extendedKeyUsage = serverAuth, clientAuth" >> $@
+	@echo "subjectAltName=@san" >> $@
+	@echo "[ san ]" >> $@
+	@echo "URI.1 = spiffe://cluster.local/ns/$(L)sa/$(SERVICE_ACCOUNT)" >> $@
+	@echo "[ req_dn ]" >> $@
+	@echo "O = $(INTERMEDIATE_ORG)" >> $@
+	@echo "CN = $(WORKLOAD_CN)" >> $@
+	@echo "L = $(L:/=)" >> $@
+
+#------------------------------------------------------------------------
+
+## <namespace>/token-cert will get a certificate for namespace using the token method. This only requires namespace
+# permissions, will not access istio-system or the certifcate.
+# CA_ADDR can be set to a different location - reflecting the gateway address
+CA_ADDR ?= istiod.istio-system.svc:15012
+# Get a token
+%/token:
+	# Using local .kube/config credentials to get a JWT token.
+	echo '{"kind":"TokenRequest","apiVersion":"authentication.k8s.io/v1","spec":{"audiences":["istio-ca"], "expirationSeconds":2592000}}' | \
+			kubectl create --raw /api/v1/namespaces/$(dir $@)serviceaccounts/default/token -f - | \
+			jq -j '.status.token' > $(dir $@)istio-token
+	# Get the Istid root CA (only for self-signed)
+	kubectl -n istio-system get secret istio-ca-secret -ojsonpath='{.data.ca-cert\.pem}' | \
```

Comment:
I think we are assuming the secret exists in istio-system ns which is not always the case.

Reply:
Also, in central istiod env, the istio-ca-secret doesn't exist in istiod cluster.  It exists in the workload cluster.

Reply:
Think the error below related to it.  
```
$ ISTIO_NAMESPACE=istio2 make --makefile ../istio/tools/certs/Makefile linistio10-certs-istioctl  
fetching root ca from k8s cluster: mycluster-wdc06-b3c.4x16-bs76p72w0qc3sa93oti0
Error from server (NotFound): secrets "istio-ca-secret" not found
Error from server (NotFound): secrets "istio-ca-secret" not found
generating linistio10/ca-key.pem
Generating RSA private key, 4096 bit long modulus
...................................++
......................................................................++
e is 65537 (0x10001)
generating linistio10/cluster-ca.csr
generating linistio10/ca-cert.pem
Signature ok
subject=/O=Istio/CN=Intermediate CA/L=linistio10
unable to load certificate
4591312364:error:09FFF06C:PEM routines:CRYPTO_internal:no start line:/BuildRoot/Library/Caches/com.apple.xbs/Sources/libressl/libressl-47.11.1/libressl-2.8/crypto/pem/pem_lib.c:684:Expecting: TRUSTED CERTIFICATE
make[1]: *** [linistio10/ca-cert.pem] Error 1
rm linistio10/cluster-ca.csr linistio10/intermediate.conf
make: *** [linistio10-certs-istioctl] Error 2
(⎈ |mycluster-wdc06-b3c.4x16/bs76p72w0qc3sa93oti0:default)
~/go/src/istio.io/testingdir ⌚ 21:51:08
$ k get secret -n istio2       
NAME                                                             TYPE                                  DATA   AGE
default-token-7dqrs                                              kubernetes.io/service-account-token   3      29d
istio-kubeconfig                                                 Opaque                                1      14d
istiod-service-account-token-td46l                               kubernetes.io/service-account-token   3      29d
mycluster-wdc06-b-406454-85f044fc29ce613c264409c04a76c95d-0001   Opaque                                2      28d
(⎈ |mycluster-wdc06-b3c.4x16/bs76p72w0qc3sa93oti0:default)
~/go/src/istio.io/testingdir ⌚ 21:51:29
$ k get pods -n istio2      
NAME                      READY   STATUS    RESTARTS   AGE
istiod-7b6565d7d5-jcbhb   1/1     Running   0          14d
```

Reply:
if i run this on the workload cluster, it doesn't work either if I use a special ns.  (I think it worked earlier when I didn't specify ns)

```
⎈ |linistio10/7caab3af9f514f028081a8180c107b69:default)
~/go/src/istio.io/testingdir ⌚ 21:54:28
$ ISTIO_NAMESPACE=istio2 make --makefile ../istio/tools/certs/Makefile linistio10-certs-istioctl 
rm: linistio10-istioctl/*: No such file or directory
fetching root ca from k8s cluster: linistio10-7caab3af9f514f028081a8180c107b69
generating linistio10/ca-key.pem
Generating RSA private key, 4096 bit long modulus
............................................................................++
........................................................................++
e is 65537 (0x10001)
generating linistio10/cluster-ca.csr
generating linistio10/ca-cert.pem
Signature ok
subject=/O=Istio/CN=Intermediate CA/L=linistio10
unable to load certificate
4732083692:error:09FFF06C:PEM routines:CRYPTO_internal:no start line:/BuildRoot/Library/Caches/com.apple.xbs/Sources/libressl/libressl-47.11.1/libressl-2.8/crypto/pem/pem_lib.c:684:Expecting: TRUSTED CERTIFICATE
make[1]: *** [linistio10/ca-cert.pem] Error 1
rm linistio10/cluster-ca.csr linistio10/intermediate.conf
make: *** [linistio10-certs-istioctl] Error 2
```

Reply:
You are right on the first part - it should fetch the ca_secrets from the current namespace. Also we can remove this step for both 'central istiod with proper DNS certs' as well as if we use --insecure on the grpcURL ( for testing ). 

But it should fetch it from the current namespace and using the configmap that istiod is creating.

It is WIP...

Reply:
Is the intention to always use the `default` service account for these short-lived tokens?